### PR TITLE
enh(ruby) Match numeric literals per language documentation

### LINKED ISSUE #2766: (ruby) numeric literals with suffixes (exponent, real, imaginary) not highligted
**Describe the issue**

In Ruby, float, rational, and complex literals do not highlight the textual portion of the literal. 

```
3.5e4
2r
6i
```

The `e4`, `r`, and `i` remain unhighlighted.

**Which language seems to have the issue?**

Ruby.

**Expected behavior**

The textual portion should also be highlighted as `number`, as the Python grammar and others do.

### PR DESCRIPTION
**Pull request by gibson042**
https://docs.ruby-lang.org/en/2.7.0/doc/syntax/literals_rdoc.html#label-Numbers

Fixes #2766

### Changes
Replaced overly-basic numeric regular expression with specific variants to match the [Ruby language documentation](https://docs.ruby-lang.org/en/2.7.0/doc/syntax/literals_rdoc.html#label-Numbers).

### Checklist
- [x] Added markup tests
- [x] Updated the changelog at `CHANGES.md`
- [x] Added myself to `AUTHORS.txt`, under Contributors

### CODE REVIEW

File: src/languages/ruby.js [hunk 1]
```diff
@@ -80,6 +80,25 @@ export default function(hljs) {
...
+  var digits = '[0-9](_?[0-9])*';
+  var NUMBER = {
+    className: 'number', relevance: 0,
+    variants: [
+      // decimal integer/float, optionally exponential or rational, optionally imaginary
+      { begin: `\\b(${decimal})(\\.(${digits}))?([eE][+-]?(${digits})|r)?i?\\b` },
+
+      // 0-prefixed octal integer or explicit decimal/binary/octal/hexadecimal integer,
+      // optionally rational and/or imaginary
+      { begin: "\\b(0[dD][0-9](_?[0-9])*|0[bB][01](_?[01])*|0([oO][0-7])?(_?[0-7])*" +
```
Comment (joshgoebel) [inline src/languages/ruby.js:97]: So we split this into individual variants for readability?  IE, one for decimal, one for binary, one for octal, one for hex, etc...

**Comment by joshgoebel** [comment]
Wait, are you all doing the same things? 

https://github.com/highlightjs/highlight.js/pull/2786
https://github.com/highlightjs/highlight.js/pull/2788

Should one be build on the other or should one of you wait?  Feels like we're duplicating effort here.

File: src/languages/ruby.js (hunk 1 above)
Comment (gibson042) [reply src/languages/ruby.js:97]: There's a performance penalty for each additional variant, and Ruby numeric literals are conveniently expressible with just one for float and one for integer. But I could split the sub-patterns into distinct lines joined by string concatenation if you prefer?

Comment (joshgoebel) [reply src/languages/ruby.js:97]: >There's a performance penalty for each additional variant  

Are you sure? :) I think not - or it's very, very slight.  Internally (inside the parse engine) it all ends up in a `|` divided regex anyways.. but in the grammar individual variants are MUCH simpler to read and understand.

Comment (gibson042) [reply src/languages/ruby.js]: OK, fair enough. Separated.

File: src/languages/ruby.js [hunk 2]
```diff
@@ -80,6 +80,30 @@ export default function(hljs) {
...
+  // Ruby syntax is underdocumented, but this grammar seems to be accurate
+  // as of version 2.7.2 (confirmed with (irb and `Ripper.sexp(...)`)
+  // https://docs.ruby-lang.org/en/2.7.0/doc/syntax/literals_rdoc.html#label-Numbers
+  var decimal = '[1-9](_?[0-9])*|0';
+  var digits = '[0-9](_?[0-9])*';
+  var NUMBER = {
+    className: 'number', relevance: 0,
+    variants: [
+      // decimal integer/float, optionally exponential or rational, optionally imaginary
+      { begin: `\\b(${decimal})(\\.(${digits}))?([eE][+-]?(${digits})|r)?i?\\b` },
```
Comment (joshgoebel) [inline src/languages/ruby.js:93]: Is the r meant to be grouped in with the exponent?

Comment (gibson042) [reply src/languages/ruby.js:93]: Yes that's the "optionally exponential or rational". An implicit decimal literal may be followed by an exponent part _or_ a rational indicator, but not both. `2.718e0` is valid and `2.718r` is valid, but `2.718e0r` is not.

File: test/markup/ruby/numbers.expect.txt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+<span class="hljs-number">1_00ri</span>; <span class="hljs-number">0_7_7ri</span>; <span class="hljs-number">4.2ri</span>; <span class="hljs-number">4.2_0ri</span>; <span class="hljs-number">88.0ri</span>; <span class="hljs-number">1_3.3_7ri</span>; <span class="hljs-number">0d8_0_0ri</span>; <span class="hljs-number">0B1_00ri</span>; <span class="hljs-number">0o70_0ri</span>; <span class="hljs-number">0XaB_cri</span>
+
+
+<span class="hljs-comment"># expressions containing numeric literals</span>
+<span class="hljs-number">0</span>..<span class="hljs-number">7</span>; <span class="hljs-number">1e1</span>..<span class="hljs-number">2e1</span>; <span class="hljs-number">0b0</span>.to_s
+
+<span class="hljs-comment"># pseudo-expressions containing numeric literals</span>
+<span class="hljs-number">0d100</span>.<span class="hljs-number">0</span>
+.<span class="hljs-number">0</span>
+<span class="hljs-number">0</span>.
```
Comment (joshgoebel) [inline test/markup/ruby/numbers.expect.txt:45]: Should those move into invalid or should we have a 3rd area for them?

Comment (joshgoebel) [reply test/markup/ruby/numbers.expect.txt:45]: n/m

**Review by joshgoebel — APPROVED** [review]

**Pull request merged**